chanfitness/chanfitness.go

```diff
@@ -0,0 +1,226 @@
+// Package chanfitness monitors the behaviour of channels to provide scores
+// which can be factored into the decision to close a channel.
+//
+// The metrics monitored are:
+// Uptime: the total time that the channel's remote peer has been observed as
+// being online by the score store.
+//
+// Lifetime: the total time the channel has been known to the scoring system.
+// Note that this does not equal channel lifetime because score data is not
+// currently persisted.
+//
+// Flap count: average the number of times the channel's remote peer has been
+// observed going offline.
+package chanfitness
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/lnpeer"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ScoreStore maintains a set of scores for the node's channels. It is intended
+// to provide insight into the performance of channels.
+type ScoreStore struct {
+	cfg *Config
+
+	channels map[uint64]*ChannelScore
+
+	quit chan struct{}
+
+	sync.RWMutex
+}
+
+// Config provides the score store with functions required to monitor peer activity.
+// All elements of the config must be non-nil for the score store to operate.
+type Config struct {
+	NotifyPeerOnline  notifyOnline
+	NotifyPeerOffline notifyOffline
+	ChannelEvents     func() (*subscribe.Client, error)
+	GetOpenChannels   func() ([]*channeldb.OpenChannel, error)
+}
+
+type (
+	notifyOnline  func(peerKey [33]byte, peerChan chan<- lnpeer.Peer)
+	notifyOffline func(peerPubKey [33]byte) <-chan struct{}
+)
+
+// NewScoreStore initializes a score store with the config provided. Note that
+// this function does not start the monitoring process, Start() must be called
+// to start the monitoring process.
+func NewScoreStore(config *Config) (*ScoreStore, error) {
+	scorer := &ScoreStore{
+		cfg:      config,
+		channels: make(map[uint64]*ChannelScore),
+		quit:     make(chan struct{}),
+	}
+
+	return scorer, nil
+}
+
+// Start adds all existing open channels to the score store and starts monitoring
+// for channel opened/closed events so that it can add and remove channels as
+// appropriate.
+func (c *ScoreStore) Start() error {
+	// Add the existing set of channels to the score store. This is required
+	// because channel events will not be triggered for channels that exist
+	// at startup time.
+	channels, err := c.cfg.GetOpenChannels()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Adding %v channels to score store", len(channels))
+
+	for _, ch := range channels {
+		var peerKey [33]byte
+		copy(peerKey[:], ch.IdentityPub.SerializeCompressed())
+
+		c.OpenChannel(ch.ShortChanID().ToUint64(), peerKey)
+	}
+
+	// channelUpdates provides a subscription to channel events which is used to
+	// listen for the creation of new channels and closing of existing channels.
+	channelUpdates, err := c.cfg.ChannelEvents()
```

Comment:
Can this be called inside `MonitorChannelEvents` instead?

Reply:
I split it up to make testing easier, because I can't set the items in a `subscribe.Client` updates queue because it's not exported. 

I originally thought that there might be a case for putting client behind and interface but I see the subscribe server is aware of the update queue structure so it's not not that simple.